src/raft/progress.rs

```diff
@@ -41,6 +41,43 @@ impl Default for ProgressState {
     }
 }
 
+/// `ProgressSet` contains several `Progress`es,
+/// which could be `Leader`, `Follower` and `Learner`.
+#[derive(Default, Clone)]
+pub struct ProgressSet {
+    pub voters: FlatMap<u64, Progress>,
+    pub learners: FlatMap<u64, Progress>,
+}
+
+impl ProgressSet {
+    pub fn new(voter_size: usize, learner_size: usize) -> Self {
+        ProgressSet {
+            voters: FlatMap::with_capacity(voter_size),
+            learners: FlatMap::with_capacity(learner_size),
+        }
+    }
+
+    pub fn get_progress(&self, id: u64) -> Option<&Progress> {
+        self.voters.get(&id).or_else(|| self.learners.get(&id))
+    }
+
+    pub fn get_mut_progress(&mut self, id: u64) -> Option<&mut Progress> {
+        let progress = self.voters.get_mut(&id);
+        if progress.is_none() {
+            return self.learners.get_mut(&id);
+        }
+        progress
+    }
+
+    pub fn check_intersect(&self) {
```

Comment:
I think this is redundant if it's checked before inserting.

Reply:
Now we didn't check before inserting.

Reply:
Why not?

Reply:
If we check it when we insert every node, we need to write the logic twice, once in `Raft::new`, and once in `set_progress`.

If we check it before inserting with `voters` and `learners` as two `Vec`, the check will be O(n^2).

So I think call `check_intersect` after the `ProgressSet` is stable is better.

Reply:
> we need to write the logic twice, once in Raft::new, and once in set_progress.

Why? What I mean is to provide an `insert` method and check in it.

> If we check it before inserting with voters and learners as two Vec, the check will be O(n^2).

How? I think it should be O(n), where n is total count of nodes and learners. Besides, usually n <= 5, so the time complexity doesn't matter here.

When define a seperate `check_intersect` method, it requires developer to remember to invoke it every time `ProgressSet` is updated, which is very easy to make error.